### ISSUE TITLE
NAN in train_loss, val_loss and Pseudo dice, 'invalid value encountered in scalar divide'

### ISSUE THREAD

**Issue by Jianghz21** [MIC-DKFZ/nnUNet#2357]
Hi, I met a problem when training nnUNetv2. When I run my dataset, I got the following outputs:

2024-07-09 13:31:29.508493: unpacking dataset...
2024-07-09 13:31:30.017582: unpacking done...
2024-07-09 13:31:30.053565: Unable to plot network architecture:
2024-07-09 13:31:30.061743: No module named 'hiddenlayer'
2024-07-09 13:31:30.147868:
2024-07-09 13:31:30.162810: Epoch 0
2024-07-09 13:31:30.168464: Current learning rate: 0.01
D:\anaconda3\Lib\site-packages\nnunetv2-2.5-py3.12.egg\nnunetv2\training\nnUNetTrainer\nnUNetTrainer.py:1119: RuntimeWarning: invalid value encountered in scalar divide
  global_dc_per_class = [i for i in [2 * i / (2 * i + j + k) for i, j, k in zip(tp, fp, fn)]]
2024-07-09 13:35:54.727426: train_loss nan
2024-07-09 13:35:54.740389: val_loss nan
2024-07-09 13:35:54.745506: Pseudo dice [nan, 0.0, 0.0]
2024-07-09 13:35:54.754245: Epoch time: 264.58 s
2024-07-09 13:35:54.762500: Yayy! New best EMA pseudo Dice: 0.0

The situation in Epoch 1, 2, ... is similar. The settings are all default. I don't know what to do next and need some help. Thanks a lot!

**Comment by seziegler** [comment]
Hi @Jianghz21 ,
usually if that happens there is something wrong with the data. Can you describe what you're trying to train and what data you have? Can you also send the dataset.json?
During the plan_and_preprocessing, did you use the `--verify_dataset_integrity` flag? If not, that would be the first thing to check.

Best,
Sebastian

**Comment by Jianghz21** [comment]
Hi, thank you for your reply!
I'm trying to train it to segment the bone defect region in CT images of the tibial plateau. I used the '--verify_dataset_integrity' flag and there was no problem. I knew there might be some problems with my data, for I only have 54 CT images in .jpg and .json, and I tried to turn them into .nii.gz. So it's quite likely that there's some problems during the conversion. The dataset.json is uploaded. I could also send the dataset to you if it would help. Could you offer me a link or an email address so I can send the dataset?
Actually, I have a new problem now. I made a few changes in the jpg-to-nii program and now I don't encounter NaN during training. But I have a new problem. The pseudo dice grows really slow  and stops growing around 0.03. Do you have any suggestions about solving this? The dataset.json that I upload corresponds to this new situation. The python code that converts .jpg and .json to .nii.gz is also uploaded (I've turned it to .txt for uploading).
Thanks a lot for your help!
[dataset.json](https://github.com/user-attachments/files/16155684/dataset.json)
[jpg_to_nii.txt](https://github.com/user-attachments/files/16155714/jpg_to_nii.txt)

**Comment by seziegler** [comment]
Hi, since you only have 2d images, are you using the 2d configuration? Can you also send the plot showing the loss and dice curves? Only 54 2d images are very little data, it might be that you're overfitting, but then your train loss should decrease normally.

**Comment by Jianghz21** [comment]
Yes, I think I'm using the 2D configuration. I upload the message I got after the preprocessing and I think that means I'm using the 2D configuration. I also upload the plot of loss and dice curves, could you please check whether that means I'm overfitting? Thanks a lot!
![2D configuration](https://github.com/MIC-DKFZ/nnUNet/assets/125629991/e203d9fd-c084-4d8f-a58a-1a30c610bf7c)
![progress](https://github.com/MIC-DKFZ/nnUNet/assets/125629991/1f014e55-cc45-465c-bf9b-cc069ca572d8)

**Comment by Jianghz21** [comment]
I got some new images yesterday and now there are 293 CT images in .jpg and .json. The loss and dice curves are as follows. Unluckily pseudo dice is still very low. Which is the more likely cause of the low dice, overfitting or mistakes in data conversion from .jpg and .json to .nii.gz? Thank you for your help~
![progress](https://github.com/MIC-DKFZ/nnUNet/assets/125629991/b5636c6f-4ec9-4e14-ba92-f463adbf5a89)

**Comment by seziegler** [comment]
Your training loss is decreasing so it looks like you're overfitting. But this can also result from some error with your data. You can check your created .nii.gz files and overlay your ground truth segmentation masks, to see if everything is still annotated as expected. Another problem could be the normalization, you used `CT` in the dataset.json but you have jpeg files. nnunet automatically uses a specific CT normalization for CT images and a z-transformation for all other image types, so you could try to use another channel_name in the dataset.json and redo the plan_and_preprocess to see if that gets you better results.

**Comment by Jianghz21** [comment]
I tried to use 3 channels "Red" "Green" and "Blue" instead and met some new problems.:

·  When I use the following code to turn the .json files into .nii.gz in labelsTr, I got this output when doing the verify_dataset_integrity：
[jpg_to_nii.txt](https://github.com/user-attachments/files/16190498/jpg_to_nii.txt)
![1720770954227](https://github.com/user-attachments/assets/cbced680-b21b-4d98-8d7a-3098aa533b68)

·  When I use the following provided in the following link (I convert my .json file to .png and mainly use the function 'convert_2d_image_to_nifti'), I got this output when doing the verify_dataset_integrity：
https://github.com/MIC-DKFZ/nnUNet/blob/nnunetv1/nnunet/dataset_conversion/Task120_Massachusetts_RoadSegm.py
https://github.com/MIC-DKFZ/nnUNet/blob/nnunetv1/nnunet/utilities/file_conversions.py
![屏幕截图 2024-07-12 160233](https://github.com/user-attachments/assets/c0a039fe-d154-43e9-845c-80b48b937a95)

·  I read on some blogs that when using 2D images, the value in mask can only be 0 and 255. So I subtract the number of labels to 2(0 for background and 255 for original, fracture and void) and then it works fairly well with dice higher than 0.86. But my task is actually to do multi-segmentation. Is it true that when dealing with 2D images there can only be 2 labels? If not, how can I produce the correct seg-files with 4 labels from .json or .png?

I'm really sorry to bother you over and over again @seziegler . But I do need some guidance to fix all this. Thank you so much for your help!

**Comment by seziegler** [comment]
Hi, no problem, I hope I can help!
You don't have to change the values for png files to 0 and 255, simply use consecutive numbers like you did before. The blogs probably meant that using 0 and 255 helps when visualizing the segmentation masks since when you just open a png image with a simple viewer you cannot really see a difference between 0 and 1, but for model training you have to use the consecutive numbers. 
Since your images are grayscale (only have 1 channel dimension) you don't need to set the channels "Red" "Green" and "Blue", simply set 
```
"channel_names": {
        "0": "myFormat"
    }
```
and it won't use CT normalization.

**Comment by Jianghz21** [comment]
Hi, I think the problem is solved ! Now I can get good results with dice around 0.9, nnUNet works pretty well with multi-segmentation task using 2D data ! 

It turns out that I did make some mistakes during the data conversion. The jpg2nii.py failed to convert .jpg files to .nii files with image size of (x, x, 1) but output .nii with shape of (512, 512) instead. I use the function `convert_2d_image_to_nifti` from the following link and set `transform = none`, now it works all right. And you're right, there is no need to use RGB channels, convert the .jpg files into grayscale and set only one channel is ok. 
https://github.com/MIC-DKFZ/nnUNet/blob/nnunetv1/nnunet/utilities/file_conversions.py

Thanks a lot for your help during the past week @seziegler !
